### ISSUE TITLE
Support a balanced image dataset generator for TripletLoss function

### ISSUE THREAD

**Issue by ma7555** [keras-team/keras-cv#97]
I have been advised by @fchollet to open the proposed solution in `keras-cv` instead of `keras`.

The problem when training a neural networ using [TripletLoss](https://github.com/tensorflow/addons/blob/8cec33fcaaf1cf90aec7bdd55a0fcdbb251ce5c2/tensorflow_addons/losses/triplet.py#L191) is that as the number of classes increases, it is highely unlikely to produce postive pairs in every batch. This results in a division by zero and thus a NaN loss and no further training happens.

The proposed data generator makes sure that a user-defined number of classes and a user-defined number of samples per class is included in every batch. If follows most of the normal `image_dataset_from_directory API`

Please see this PR in Keras Core: https://github.com/keras-team/keras/pull/15990

**Comment by LukeWood** [comment]
Thanks for offering to contribute this!

I agree that this fits in the KerasCV library - alongside triplet loss.  We'll want to provide a coherent API for SSL pretraining.  This will likely need to be part of it.  

As for the actual API and code:  I agree with Francois comments labelled 1.) and 2.) in the other PR.  

> The problem when training a neural networ using [TripletLoss](https://github.com/tensorflow/addons/blob/8cec33fcaaf1cf90aec7bdd55a0fcdbb251ce5c2/tensorflow_addons/losses/triplet.py#L191) is that as the number of classes increases, it is highely unlikely to produce postive pairs in every batch. This results in a division by zero and thus a NaN loss and no further training happens.

Is there a way to work around this in triplet loss itself?  i.e. divide_no_nan and return zero gradient for that batch?  We will be offering our own triplet loss so we can possibly work around this.

**Comment by ma7555** [comment]
Hello @LukeWood,

So regarding your question of working around this in the loss function itself. I don't think it is a good solution as this will add a major convergence drawback due to the fact that most of the batches will yield no gradient:

* When working with a large number of classes, it is highly unlikely that any of the batches would contain triplets at all. You would be just wasting resources training on a network that will never converge.

* In my case, I am training a network with ~9000 class, you can see the math below as it was made by [HauserA](https://github.com/HauserA)
 in https://github.com/tensorflow/addons/issues/1676 in this [colab notebook ](https://colab.research.google.com/drive/1qPjOIu5Dr9ADQmVfn7SXo-MPGSt8WT9x)
```
import math
f = math.factorial
def permutations(M,j):
  return (f(M) / (f(j) * f(M - j)))
def triplets(M,T,j):
  return permutations(M,j) * ( (T-1)/T )**(M-j) * (1/T)**j
def with_probability(batch_size, num_classes):
  M = batch_size
  T = num_classes
  s = 0
  for j in range(2, M):
    s+= triplets(M,T,j) * j * T
  return s
num_classes = 9000
print("Batch_size \t with_probability")
for batch_size in (16,32,64,128):
  print(f"{batch_size} \t\t {with_probability(batch_size,num_classes)}")
```
```
Batch_size 	 probability
16 		 0.02664593590888058
32 		 0.11003871567653176
64 		 0.446460369374086
128 		 1.7936370005754705
256 		 7.151933146144015
```
* I can fit 32 samples per batch on my GPU, so returning 0 gradients will just make me ignore 89% of the batches during training.
* In order for the triplet loss function to use all the samples in the batch, there must be at least 2 images of of every class.
* For example, when having a `batch_size=32`, you should use P=8, K=4. Where P is number of classes per batch and K is number of samples per class.
* So just masking the problem in the loss function is creating a hard to debug problem and a waste of time and resources.

Regarding the other points raised by @fchollet

1. The logic and arguments expected are different. They might look like each other in some parts but there are a lot of differences that would just make it easier to write it again. For example, this generator does not expect batch_size.
2. The API should be pretty much similar to the current one introduced by [image_dataset_from_directory](https://github.com/keras-team/keras/blob/v2.8.0/keras/preprocessing/image_dataset.py#L30-L235)

If you want to reuse image_dataset_from_directory function API, there will need to be 3 new arguments introducted to it (P, K and safe_triplet)

**Comment by chjort** [comment]
There is a project called [Tensorflow Similarity](https://github.com/tensorflow/similarity), which specifically provides an API for metric learning and self-supervised learning.

They have a module with different data samplers, designed to sample classes properly for these types of losses. Specifically they have the [MultiShotMemorySampler](https://github.com/tensorflow/similarity/blob/f3ad15199de08f16aca318e51771f256dc5832a0/tensorflow_similarity/samplers/memory_samplers.py#L30) which could be used to solve your problem. They have a [Hello World notebook](https://github.com/tensorflow/similarity/blob/master/examples/supervised_hello_world.ipynb) where they use this sampler.

These samplers subclass from `tf.keras.utils.Sequence`, so they can be used in any keras workflow.

**Comment by ma7555** [comment]
@chjort 
Thanks for pointing out the [Tensorflow Similarity](https://github.com/tensorflow/similarity) project. The [MultiShotMemorySampler](https://github.com/tensorflow/similarity/blob/f3ad15199de08f16aca318e51771f256dc5832a0/tensorflow_similarity/samplers/memory_samplers.py#L30) indeed can be used. However, it expects `(x, y)` I beleive in form of tensors as called by [select_examples](https://github.com/tensorflow/similarity/blob/f3ad15199de08f16aca318e51771f256dc5832a0/tensorflow_similarity/samplers/utils.py#L25). I beleive this is not a generator that reads data from disk batch by batch and it needs the full data to be stored in memory (hence its name).

I guess in order to use this, you would need to pass the x argument as file paths from [tf.data.Dataset.list_files](https://github.com/tensorflow/tensorflow/blob/v2.8.0/tensorflow/python/data/ops/dataset_ops.py#L1292-L1362) then fully extract the labels from those paths (y) then map after batching with a function that would take a batch and return the decoded images and their labels.

I find this workflow should be reserved for advanced usage. The `image_dataset_from_directory` API is much simpler

**Comment by bhack** [comment]
The not full in memory version Is:

https://github.com/tensorflow/similarity/blob/master/tensorflow_similarity/samplers/tfrecords_samplers.py#L22

**Comment by ma7555** [comment]
Using TFRecords shards is not very common in CV datasets. Creating one out of a directory of images is not worth it too as you would most likely follow the same pipeline in the previous comment. You need to read images, extract labels and then encode and save as .tfrec

**Comment by LukeWood** [comment]
> There is a project called [Tensorflow Similarity](https://github.com/tensorflow/similarity), which specifically provides an API for metric learning and self-supervised learning.
> 
> They have a module with different data samplers, designed to sample classes properly for these types of losses. Specifically they have the [MultiShotMemorySampler](https://github.com/tensorflow/similarity/blob/f3ad15199de08f16aca318e51771f256dc5832a0/tensorflow_similarity/samplers/memory_samplers.py#L30) which could be used to solve your problem. They have a [Hello World notebook](https://github.com/tensorflow/similarity/blob/master/examples/supervised_hello_world.ipynb) where they use this sampler.
> 
> These samplers subclass from `tf.keras.utils.Sequence`, so they can be used in any keras workflow.
> 
> @LukeWood How will KerasCV position itself in relation to TF Similarity regarding a SSL API?

Yeah, tf similarity is happy to user KerasCV apis as we create them.  They’re going to adopt our CutMix and RandAugment once I finish it.

**Comment by LukeWood** [comment]
If this functionality exists in TensorFlow similarity already as @bhack has pointed out maybe it is best to leave it there for the time being.  Once we breach SSL we can reconsider.

thoughts?

**Comment by ma7555** [comment]
@LukeWood, I agree the [MultiShotMemorySampler](https://github.com/tensorflow/similarity/blob/f3ad15199de08f16aca318e51771f256dc5832a0/tensorflow_similarity/samplers/memory_samplers.py#L30) is a way to do it along with [tf.data.Dataset.list_files](https://github.com/tensorflow/tensorflow/blob/v2.8.0/tensorflow/python/data/ops/dataset_ops.py#L1292-L1362) while following a similar approach as this one [using_tfdata_for_finer_control](https://www.tensorflow.org/tutorials/load_data/images#using_tfdata_for_finer_control). I don't see using TFRecords shards as a feasible solution at all as almost noone uses them now after tf.data.Dataset has been out and you simply won't find datasets stored in this format anymore.

Meanwhile, I have pushed the proposed solution directly from image files in directories as a separate library for future readers who might need it out of the box here [kerasgen](https://github.com/ma7555/kerasgen)

Another point, since you mentioned you are working on some augmentations (CutMix and RandAugment). I remember I have implemented 3 augmentations 2 years ago to use with Keras as well.

* Augmix: https://arxiv.org/abs/1912.02781
* Gridmask: https://arxiv.org/abs/2001.04086
* Morphological (erode, dilate, open, close): https://opencv-python-tutroals.readthedocs.io/en/latest/py_tutorials/py_imgproc/py_morphological_ops/py_morphological_ops.html

Please feel free to visit https://github.com/ma7555/Augz and copy the logic you might need if you plan to implement any of them or maybe I can open a PR for them as well if they are needed.

**Comment by ma7555** [comment]
An example of the full TripletLoss workflow: [Colab](https://colab.research.google.com/github/ma7555/kerasgen/blob/master/examples/triplet_loss.ipynb).

I guess I will close for now. If you feel this API is needed please reopen.
Thanks!
